### DISCUSSION TITLE
Delete nodes and edges in the version 10.

### DISCUSSION THREAD

**Discussion by JazielIsai** [xyflow/xyflow#1990] [Q&A] [answered]
Good morning, I have some questions about react-flow in your version 10.

How I can delete nodes and edges without using the backspace?
Which function I can using for delete nodes and edges and which is your structure or syntax?

I await your response
Thank you very much.

**Comment by anto-belo** [comment]
Maybe it's not documented yet, but yesterday I found this by debugging what happens after backspace is pressed and onEdgesDelete event is activated, so it worked for me:
`setEdges((eds) => applyEdgeChanges([
                        {
                            id: nodeId,
                            type: 'remove'
                        },
                        {
                            id: nodeId,
                            type: 'select',
                            selected: false
                        }
                    ], eds))`

**Comment by JazielIsai** [comment]
Thank you very much

**Comment by moklick** [comment, marked as answer]
Hey @JazielIsai 

if you want to remove a node or an edge, you need to remove that specific element from the nodes/edges array. There are multiple ways to do this but you don't need to use `applyNodeChanges` or `applyEdgeChanges`. 

You could do, for example:

```js
const deleteNodeById = (id) => {
  setNodes(nds => nds.filter(node => node.id !== id));
};

**Reply by JazielIsai** [reply]
Thank you very much!!

**Reply by VictorAITERA** [reply]
Thanks for the answer but the problem is that onNodesDelete is not called after doing that on ReactFlow component.
Something that I do wrong ?

**Reply by KaiyuanMa** [reply]
> Thanks for the answer but the problem is that onNodesDelete is not called after doing that on ReactFlow component. Something that I do wrong ?

Yeah the same problem happened to me. I think noNodesDelete only triggers when you press "backSpace". It does not trigger if you do a filter or even pass onNodesChange(manualChange) to your custom delete function.

This will cause you some issues including: 
cannot reuse noNodesDelete,
your custom delete function does not trigger onEdgesChange which means the edges that are connected to the deleted node are still in your edges state, if you want to delete the edges, you will to use getConnectedEdges() to find all the edges and delete them in your delete function. However, getConnectedEdges() wants you pass in your target nodes and all the edges, which means if your delete function is in your react child component, you need to pass in the edges state and update it everytime or you need to declare the delete function in your ReactFlow component. BUT I when I declare the function in my ReactFlow component, edges are still empty even though its not, this might be solved by using callback, but I haven't figure it out yet. Some help would be great.

**Reply by afreenaffu-web** [reply]
Did you figured it out ? even i am facing the same problem.

**Reply by AKiranB** [reply]
It can be a bit tricky to make everything work on custom node delete.
If you have ReactFlow correctly wrapped in the provider you should be able to access the current edges using useReactFlow()

`reactFlowInstance = useReactFlow()
const currentEdges = reactFlowInstance.getEdges()
`

Using this, as mentioned by @KaiyuanMa, you should be able to use getConnectedEdges() and eventually simply update the edges array using setEdges

` const connectedEdges = getConnectedEdges(node, edges)
    setEdges(edges => [...edges.filter(edge => !connectedEdges.includes(edge))])`
    
   This has so far worked for me, and I hope it helps a little bit.
